### ISSUE TITLE
Function parameter names in Typescript

### ISSUE THREAD

**Issue by Yakimych** [cristianoc/genType#155]
Hi. I am trying to use genType in order to be able to use Reason/OCaml functions in an existing TypeScript project, but I can't seem to get it to generate functions with the same input parameter names - instead I get `_1`, `_2`, etc. Is there something simple I am missing?

Here is a simple function in Reason:
```
[@genType]
let sum = (a: int, b: int): int => a + b;
```

The JS version comes with the same parameter names `a` and `b`:
```
function sum(a, b) {
  return a + b | 0;
}
```

The TypeScript version, however, looks as follows:
```
export const sum: (_1:number, _2:number) => number = function _(Arg1: any, Arg2: any) {
  const result = Curry._2(somefuncsBS.sum, Arg1, Arg2);
  return result
};
```

Is there any way to generate `a` and `b` as parameter names in TypeScript instead of `_1` and `_2`? (Looks like the code that generates those names is here: https://github.com/cristianoc/genType/blob/master/src/EmitType.re#L264)

The closest I've come is by using named parameters `~a` and `~b`, but then I have to pass an object when calling the function from TS: `sum({ a: 1, b: 2})`;

**Comment by cristianoc** [comment]
@Yakimych could you explain the concern here? Is the goal to generate more readable code, or is there something else as well?

**Comment by cristianoc** [comment]
The code is generated starting from types, and the AST of types only contains names for labeled arguments:
```ocaml
type arg_label =
    Nolabel
  | Labelled of string (*  label:T -> ... *)
  | Optional of string (* ?label:T -> ... *)
```.

So I guess there's not much that can be done.

**Comment by Yakimych** [comment]
@cristianoc not so much of a "concern", just investigating possibilities of using Reason/OCaml code from an existing TypeScript codebase.

It's not so much about "readable" code as in I am not really going to read it, but this is information that helps to ***use*** the code. The problem is that it's rather difficult to figure out what a function like this expects (sure I can see the types, but not the variable names). With JS I can at least go to definition and see the names of the parameters (e.g. `id`, `name`). With the TS version the intellisense just shows `_1`, `_2`, which doesn't tell anything to the function caller. Going to definition doesn't help either, since I don't get any more information from there.

In other words, it requires jumping through a lot of hoops in order to find either the JS-code or Reason source code and figure out what parameters the function expects, which kind partly kills the point of using TypeScript (at least one the points). But I suppose if it's not technically possible, then there isn't anything to be done.

One though/question - is there a way to make this work with `[@genType.as "name"]` annotations for function arguments?

**Comment by cristianoc** [comment]
@Yakimych  thank you. Technically, it's not feasible right now with genType starting from types as input. Theoretically, if one had deeper compiler integration, then one could do more.

>One though/question - is there a way to make this work with `[@genType.as "name"]` annotations for function arguments?

It would be possible to add support for user-provided annotations to determine what would appear in the generated code instead of `_1` and `_2`.

**Comment by Yakimych** [comment]
@cristianoc ok, thanks!

I am still relatively new to Reason/OCaml, but I will have some time next month to dig deeper into the codebase and see if I can contribute.

**Comment by cristianoc** [comment]
Because of the issue mentioned here https://github.com/cristianoc/genType#functions, the first parameter of a function cannot be annotated anyway, until the switch to compiler 4.06 happens.

Best to look into this after 4.06.

**Comment by cristianoc** [comment]
4.06 is out.
